README.md
nagios cookbook
===============

[![Join the chat at https://gitter.im/schubergphilis/nagios](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/schubergphilis/nagios?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Build Status](https://travis-ci.org/schubergphilis/nagios.svg)](https://travis-ci.org/schubergphilis/nagios)
[![Cookbook Version](https://img.shields.io/cookbook/v/nagios.svg)](https://supermarket.chef.io/cookbooks/nagios)

Installs and configures Nagios server. Chef nodes are automatically discovered using search, and Nagios host groups are created based on Chef roles and optionally environments as well.


Requirements
------------
### Chef
Chef version 0.10.10+ and Ohai 0.6.12+ are required.

Because of the heavy use of search, this recipe will not work with Chef Solo, as it cannot do any searches without a server.

This cookbook relies heavily on multiple data bags. See __Data Bag__ below.

The system running this cookbooks should have a role named 'monitoring' so that NRPE clients can authorize monitoring from that system. This role name is configurable via an attribute. See __Attributes__ below.

The functionality that was previously in the nagios::client recipe has been moved to its own NRPE cookbook at https://github.com/schubergphilis/nrpe

### Platform
* Debian 6.X, 7.X
* Ubuntu 10.04, 12.04, 13.04
* Red Hat Enterprise Linux (CentOS/Amazon/Scientific/Oracle) 5.X, 6.X

**Notes**: This cookbook has been tested on the listed platforms. It may work on other platforms with or without modification.

### Cookbooks
* apache2 2.0 or greater
* build-essential
* nginx
* nginx_simplecgi
* php
* yum-epel (note: this requires yum cookbook v3.0, which breaks compatibility with many other cookbooks)


Attributes
----------
### config
[The config file](https://github.com/schubergphilis/nagios/blob/master/attributes/config.rb) contains the Nagios configuration options. Consult the [nagios documentation](http://nagios.sourceforge.net/docs/3_0/configmain.html) for available settings and allowed options. Configuration entries of which multiple entries are allowed, need to be specified as an Array.

Example: `default['nagios']['conf']['cfg_dir'] = [ '/etc/nagios/conf.d' , '/usr/local/nagios/conf.d' ]`

### default
* `node['nagios']['user']` - Nagios user, default 'nagios'.
* `node['nagios']['group']` - Nagios group, default 'nagios'.
* `node['nagios']['plugin_dir']` - location where Nagios plugins go, default '/usr/lib/nagios/plugins'.
* `node['nagios']['multi_environment_monitoring']` - Chef server will monitor hosts in all environments, not just its own, default 'false'
* `node['nagios']['monitored_environments']` - If multi_environment_monitoring is 'true' nagios will monitor nodes in all environments. If monitored_environments is defined then nagios will monitor only hosts in the list of environments defined. For ex: ['prod', 'beta'] will monitor only hosts in 'prod' and 'beta' chef_environments. Defaults to '[]' - and all chef environments will be monitored by default.
* `node['nagios']['monitoring_interface']` - If set, will use the specified interface for all nagios monitoring network traffic. Defaults to `nil`
* `node['nagios']['exclude_tag_host']` - If set, hosts tagged with this value will be excluded from nagios monitoring.  Defaults to ''

* `node['nagios']['server']['install_method']` - whether to install from package or source. Default chosen by platform based on known packages available for Nagios: debian/ubuntu 'package', redhat/centos/fedora/scientific: source
* `node['nagios']['server']['install_yum-epel']` - whether to install the EPEL repo or not (only applies to RHEL platform family). The default value is `true`. Set this to `false` if you do not wish to install the EPEL RPM; in this scenario you will need to make the relevant packages available via another method e.g. local repo, or install from source.
* `node['nagios']['server']['service_name']` - name of the service used for Nagios, default chosen by platform, debian/ubuntu "nagios3", redhat family "nagios", all others, "nagios"
* `node['nagios']['home']` - Nagios main home directory, default "/usr/lib/nagios3"
* `node['nagios']['conf_dir']` - location where main Nagios config lives, default "/etc/nagios3"
* `node['nagios']['resource_dir']` - location for recources, default "/etc/nagios3"
* `node['nagios']['config_dir']` - location where included configuration files live, default "/etc/nagios3/conf.d"
* `node['nagios']['log_dir']` - location of Nagios logs, default "/var/log/nagios3"
* `node['nagios']['cache_dir']` - location of cached data, default "/var/cache/nagios3"
* `node['nagios']['state_dir']` - Nagios runtime state information, default "/var/lib/nagios3"
* `node['nagios']['run_dir']` - where pidfiles are stored, default "/var/run/nagios3"
* `node['nagios']['docroot']` - Nagios webui docroot, default "/usr/share/nagios3/htdocs"
* `node['nagios']['timezone']` - Nagios timezone, defaults to UTC
* `node['nagios']['enable_ssl]` - boolean for whether Nagios web server should be https, default false
* `node['nagios']['ssl_cert_file']` = Location of SSL Certificate File. default "/etc/nagios3/certificates/nagios-server.pem"
* `node['nagios']['ssl_cert_chain_file']` = Optional location of SSL Intermediate Certificate File. No default.
* `node['nagios']['ssl_cert_key']`  = Location of SSL Certificate Key. default "/etc/nagios3/certificates/nagios-server.pem"
* `node['nagios']['http_port']` - port that the Apache/Nginx virtual site should listen on, determined whether ssl is enabled (443 if so, otherwise 80). Note:  You will also need to configure the listening port for either NGINX or Apache within those cookbooks.
* `node['nagios']['server_name']` - common name to use in a server cert, default "nagios"
* `node['nagios']['server']['server_alias']` - alias name for the webserver for use with Apache.  Defaults to nil
* `node['nagios']['ssl_req']` - info to use in a cert, default `/C=US/ST=Several/L=Locality/O=Example/OU=Operations/CN=#{node['nagios']['server_name']}/emailAddress=ops@#{node['nagios']['server_name']}`

*  `node['nagios']['server']['url']` - url to download the server source from if installing from source
*  `node['nagios']['server']['version']` - version of the server source to download
*  `node['nagios']['server']['checksum']` - checksum of the source files
*  `node['nagios']['server']['patch_url']` - url to download patches from if installing from source
*  `node['nagios']['server']['patches']` - array of patch filenames to apply if installing from source
*  `node['nagios']['url']` - URL to host Nagios from - defaults to nil and instead uses  FQDN

* `node['nagios']['conf']['enable_notifications']` - set to 1 to enable notification.
* `node['nagios']['conf']['interval_length']` - minimum interval. Defaults to '1'.

* `node['nagios']['check_external_commands']`
* `node['nagios']['default_contact_groups']`
* `node['nagios']['sysadmin_email']` - default notification email.
* `node['nagios']['sysadmin_sms_email']` - default notification sms.
* `node['nagios']['server_auth_method']` - authentication with the server can be done with openid (using `apache2::mod_auth_openid`), cas (using `apache2::mod_auth_cas`),ldap (using `apache2::mod_authnz_ldap`), or htauth (basic). The default is htauth. "openid" will utilize openid authentication, "cas" will utilize cas authentication, "ldap" will utilize LDAP authentication, and any other value will use htauth (basic).
* `node['nagios']['cas_login_url']` - login url for cas if using cas authentication.
* `node['nagios']['cas_validate_url']` - validation url for cas if using cas authentication.
* `node['nagios']['cas_validate_server']` - whether to validate the server cert. Defaults to off.
* `node['nagios']['cas_root_proxy_url']` - if set, sets the url that the cas server redirects to after auth.
* `node['nagios']['ldap_bind_dn']` - DN used to bind to the server when searching for ldap entries.
* `node['nagios']['ldap_bind_password']` - bind password used with the DN provided for searching ldap.
* `node['nagios']['ldap_url']` - ldap url and search parameters.
* `node['nagios']['ldap_authoritative']` - accepts "on" or "off". controls other authentication modules from authenticating the user if this one fails.
* `node['nagios']['users_databag']` - the databag containing users to search for. defaults to users
* `node['nagios']['users_databag_group']` - users databag group considered Nagios admins.  defaults to sysadmin
* `node['nagios']['services_databag']` - the databag containing services to search for. defaults to nagios_services
* `node['nagios']['servicegroups_databag']` - the databag containing servicegroups to search for. defaults to nagios_servicegroups
* `node['nagios']['templates_databag']` - the databag containing templates to search for. defaults to nagios_templates
* `node['nagios']['hosttemplates_databag']` - the databag containing host templates to search for. defaults to nagios_hosttemplates
* `node['nagios']['eventhandlers_databag']` - the databag containing eventhandlers to search for. defaults to nagios_eventhandlers
* `node['nagios']['unmanaged_hosts_databag']` - the databag containing unmanagedhosts to search for. defaults to nagios_unmanagedhosts
* `node['nagios']['serviceescalations_databag']` - the databag containing serviceescalations to search for. defaults to nagios_serviceescalations
* `node['nagios']['hostescalations_databag']` - the databag containing hostescalations to search for. defaults to nagios_hostescalations
* `node['nagios']['contacts_databag']` - the databag containing contacts to search for. defaults to nagios_contacts
* `node['nagios']['contactgroups_databag']` - the databag containing contactgroups to search for. defaults to nagios_contactgroups
* `node['nagios']['servicedependencies_databag']` - the databag containing servicedependencies to search for. defaults to nagios_servicedependencies
* `node['nagios']['host_name_attribute']` - node attribute to use for naming the host. Must be unique across monitored nodes. Defaults to hostname
* `node['nagios']['regexp_matching']` - Attribute to enable [regexp matching](http://nagios.sourceforge.net/docs/3_0/configmain.html#use_regexp_matching). Defaults to 0.
* `node['nagios']['large_installation_tweaks']` - Attribute to enable [large installation tweaks](http://nagios.sourceforge.net/docs/3_0/largeinstalltweaks.html). Defaults to 0.
* `node['nagios']['templates']` - These set directives in the default host template. Unless explicitly overridden, they will be inherited by the host definitions for each discovered node and `nagios_unmanagedhosts` data bag. For more information about these directives, see the Nagios documentation for [host definitions](http://nagios.sourceforge.net/docs/3_0/objectdefinitions.html#host).
* `node['nagios']['hosts_template']` - Host template you want to inherit properties/variables from, default 'server'. For more information, see the nagios doc on [Object Inheritance](http://nagios.sourceforge.net/docs/3_0/objectinheritance.html).
* `node['nagios']['brokers']` - Hash of broker modules to include in the config. Hash key is the path to the broker module, the value is any parameters to pass to it.


* `node['nagios']['default_host']['flap_detection']` - Defaults to `true`.
* `node['nagios']['default_host']['process_perf_data']` - Defaults to `false`.
* `node['nagios']['default_host']['check_period']` - Defaults to `'24x7'`.
* `node['nagios']['default_host']['check_interval']` - In seconds. Must be divisible by `node['nagios']['interval_length']`. Defaults to `15`.
* `node['nagios']['default_host']['retry_interval']` - In seconds. Must be divisible by `node['nagios']['interval_length']`. Defaults to `15`.
* `node['nagios']['default_host']['max_check_attempts']` - Defaults to `1`.
* `node['nagios']['default_host']['check_command']` - Defaults to the pre-defined command `'check-host-alive'`.
* `node['nagios']['default_host']['notification_interval']` - In seconds. Must be divisible by `node['nagios']['interval_length']`. Defaults to `300`.
* `node['nagios']['default_host']['notification_options']` - Defaults to `'d,u,r'`.
* `node['nagios']['default_host']['action_url']` - Defines a action url.  Defaults to `nil`.

* `node['nagios']['default_service']['process_perf_data']` - Defaults to `false`.
* `node['nagios']['default_service']['action_url']` - Defines a action url. Defaults to `nil`.

* `node['nagios']['server']['web_server']` - web server to use. supports Apache or Nginx, default "apache"
* `node['nagios']['server']['nginx_dispatch']` - nginx dispatch method. supports cgi or php, default "cgi"
* `node['nagios']['server']['stop_apache']` - stop apache service if using nginx, default false
* `node['nagios']['server']['redirect_root']` - if using Apache, should http://server/ redirect to http://server/nagios3 automatically, default false
* `node['nagios']['server']['normalize_hostname']` - If set to true, normalize all hostnames in hosts.cfg to lowercase. Defaults to false.

 These are nagios cgi.config options.

 * `node['nagios']['cgi']['show_context_help']`                         - Defaults to 1
 * `node['nagios']['cgi']['authorized_for_system_information']`         - Defaults to '*'
 * `node['nagios']['cgi']['authorized_for_configuration_information']`  - Defaults to '*'
 * `node['nagios']['cgi']['authorized_for_system_commands']`            - Defaults to '*'
 * `node['nagios']['cgi']['authorized_for_all_services']`               - Defaults to '*'
 * `node['nagios']['cgi']['authorized_for_all_hosts']`                  - Defaults to '*'
 * `node['nagios']['cgi']['authorized_for_all_service_commands']`       - Defaults to '*'
 * `node['nagios']['cgi']['authorized_for_all_host_commands']`          - Defaults to '*'
 * `node['nagios']['cgi']['default_statusmap_layout']`                  - Defaults to 5
 * `node['nagios']['cgi']['default_statuswrl_layout']`                  - Defaults to 4
 * `node['nagios']['cgi']['escape_html_tags']`                          - Defaults to 0
 * `node['nagios']['cgi']['action_url_target']`                         - Defaults to '_blank'
 * `node['nagios']['cgi']['notes_url_target']`                          - Defaults to '_blank'
 * `node['nagios']['cgi']['lock_author_names']`                         - Defaults to 1


* `node['nagios']['nodes_search_query']` - The search query for finding all the nodes you want the nagios server to check. Defaults to "hostname:* AND chef_environment:#{node.chef_environment}". However, if you set the node['nagios']['multi_environment_monitoring'] to true, it will always be "hostname:*".

Recipes
-------
### default
Includes the correct client installation recipe based on platform, either `nagios::server_package` or `nagios::server_source`.

The server recipe sets up Apache as the web front end by default. This recipe also does a number of searches to dynamically build the hostgroups to monitor, hosts that belong to them and admins to notify of events/alerts.

Searches are confined to the node's `chef_environment` unless multi-environment monitoring is enabled.

The recipe does the following:

1. Searches for users in 'users' databag belonging to a 'sysadmin' group, and authorizes them to access the Nagios web UI and also to receive notification e-mails.
2. Searches all available roles/environments and builds a list which will become the Nagios hostgroups.
3. Places nodes in Nagios hostgroups by role / environment membership.
4. Installs various packages required for the server.
5. Sets up configuration directories.
6. Moves the package-installed Nagios configuration to a 'dist' directory.
7. Disables the 000-default VirtualHost present on Debian/Ubuntu Apache2 package installations.
8. Templates configuration files for services, contacts, contact groups, templates, hostgroups and hosts.
9. Enables the Nagios web UI.
10. Starts the Nagios server service


### server\_package
Installs the Nagios server from packages. Default for Debian / Ubuntu systems.

### server\_source
Installs the Nagios server from source. Default for Red Hat / Fedora based systems as native packages for Nagios are not available in the default repositories.

### pagerduty
Installs pagerduty plugin for nagios. If you only have a single pagerduty key, you can simply set a `node['nagios']['pagerduty_key']` attribute on your server.  For multiple pagerduty key configuration see Pager Duty under Data Bags.

This recipe was written based on the [Nagios Integration Guide](http://www.pagerduty.com/docs/guides/nagios-integration-guide) from PagerDuty which explains how to get an API key for your Nagios server.


Data Bags
---------
[See Wiki for more databag information](https://github.com/schubergphilis/nagios/wiki/config)

### Pager Duty
You can define pagerduty contacts and keys by creating nagios\_pagerduty data bags that contain the contact and
the relevant key. Setting admin\_contactgroup to "true" will add this pagerduty contact to the admin contact group
created by this cookbook.

```javascript
{
  "id": "pagerduty_critical",
  "admin_contactgroup": "true",
  "key": "a33e5ef0ac96772fbd771ddcccd3ccd0"
}
```

You can add these contacts to any contactgroups you create.

Monitoring Role
---------------
Create a role to use for the monitoring server. The role name should match the value of the attribute "`node['nrpe']['server_role']`" on your clients. By default, this is '`monitoring`'. For example:

```ruby
# roles/monitoring.rb
name 'monitoring'
description 'Monitoring server'
run_list(
  'recipe[nagios::default]'
)

default_attributes(
  'nagios' => {
    'server_auth_method' => 'htauth'
  }
)
```

```bash
$ knife role from file monitoring.rb
```

Usage
-----
### server setup
Create a role named '`monitoring`', and add the nagios server recipe to the `run_list`. See __Monitoring Role__ above for an example.

Apply the nrpe cookbook to nodes in order to install the NRPE client

By default the Nagios server will only monitor systems in its same environment. To change this set the `multi_environment_monitoring` attribute. See __Attributes__

Create data bag items in the `users` data bag for each administer you would like to be able to login to the Nagios server UI. Pay special attention to the method you would like to use to authorization users (openid or htauth). See __Users__ and __Atttributes__


CHANGELOG.md
nagios Cookbook CHANGELOG
=========================
This file is used to list changes made in each version of the nagios cookbook.
7.2.99
-----
### Improvement
- Enabled performance metrics
- Added custom node search ability

7.2.2
-----
### Bug
- Fixing the apache mpm breaking on centos.

7.2.0
-----
### Testing
- Added centos 7.1 for testing.
- Added centos 5.11 for testing.
- Added test-kitchen tests.

### Improvement
- Added logic to exclude nodes based on tag.
- Including apache2::mpm_prefork for apache.
- Added the ability to specify command arguments within services.
- Added the ability to specify custom options on hosts, contacts and services.

7.1.8
-----
### Bug
- Fixing the unmanagedhosts databag filter on environment.
- Fixing the services databag filter on environment.

### Improvement
- Moving the LWRP's providers into definitions.
  This will remove some extra complexity and output will be
  much nicer and debugging will be easier during the chef-converge.

7.1.6
-----
### Bug
- Fixing the nagios_[resource] provider delete action.

### Improvement
- Added option for custom apache auth based on attribute.
- Update cgi-path attibute on source install.
- Update on test-kitchen tests.
- Update on kitchen-vagrant version.

7.1.4
-----
### Bug
- AuthzLDAPAuthoritative is removed in Apache 2.4.
- Fixed the pagerduty config by using LWRP.

### Improvement
- Made test config os (in)dependent.
- Added zap for config file cleanup.
- Added encrypted user databag support.
- Added extra configuration tests.
- Added gitter badge.

7.1.2
-----
### Bug
- Fixed display of style sheets on Ubuntu 14.04+
- service_check_timeout_state config option is now only set on modern Nagios releases.  This broke Ubuntu 10.04/12.04 service startup
- Updated Test Kitchen release / added additional platforms for testing
- Fixed the attribute used to enable notifications in the Readme file
- Fixed loading of node['nagios']['host_name_attribute']

### Improvement
- Search queries in hostgroups data bag are now limited to the monitored environments if using node['nagios']['monitored_environments']

7.1.0
-----
### Bug
- Fixed class-type checking with duck-typing on update_options.
- Fixed host_name_attribute on nagios model.

### Improvement
- Moved all nagios configuration options within attributes.
- Moved all nagios configuration attributes into separate file.

### Breaking Changes
- With the change above we might introduced some config problems.
  Please check your attributes when upgrading.

### Development
- Added extra kitchen serverspec tests.

7.0.8
-----
### Bug
- Fixed servicegroups members.
- Chaned the order of data bag loading (commands first).

### Improvement
- Cleanup of the internals of the nagios model.

### Development
- Added kitchen serverspec tests.

7.0.6
-----
### Bug
- Fixed data bag import.(#346)
- Fixed missing create method on Servicegroup object. (#348)
- Fixed update_dependency_members for depedency objects.

7.0.4
-----
### Bug
- Fixed the order for resource.cfg population to be correct.

7.0.2
-----
### Bug
- Fixed the hardcoded cgi-bin path in server source.
- Fixed contact_groups within load_default_config recipe.
- Removed dead code from timeperiod.rb library.
- Ignore timeperiods that don't comply.
- Making time formats less restrictive. (#336)

### Improvement
- Make yum-epel recipe include optional via attribute.
- Only allow_empty_hostgroup_assignment for Nagios versions >= 3.4.0

7.0.0
-----
### Feature
- Added providers for all nagios configuration objects.
- Added wiki pages explaining the providers.
- Added wiki pages explaining the databags.

### Development
- Updated chefspec (4.2.0)

### Extra note
- Please test this version before using it in production. Some logic and attributes have changes, so this might break your current setup.

6.1.2
----------
### Feature
- Allow defining parents in the unmanaged hosts data bag so you can build the host map.

### Bug
- Setup Apache2 before trying to configure the webserver so paths will be created
- Installed EPEL on RHEL so package installs work
- Set the Apache log dir to that provided by Apache since the Nagios log dir is now locked down to just the nagios user / group
- Template the resource.cfg file on RHEL platforms to prevent check failures
- Fix cgi-bin page loads on RHEL systems
- Fix CSS files not loading on Debian based systems

### Development
- Updated Test Kitchen dependency to 1.3.1 from 1.2.1

6.1.0
-----

### Bug
- Fix missing CSS files on RHEL/Fedora package installs
- Ensure the source file for Nagios is always downloaded to work around corrupt partial downloads
- Fixed permissions being changed on the resource directory during each run on RHEL systems

### Improvement
- Remove support for SSL V2 / V3 (Apache2/NGINX) and add TLS 1.1 and 1.2 (NGINX)
- Cleaned up and removed duplicate code from the web server configuration

### New Features
- Added the ability to tag nodes with an attribute that excludes them from the monitoring search.  See readme for details

### Breaking Changes
- The /nagios or /nagios3 URLs are no longer valid.  Nagios should be installed on the root of the webserver and this never entirely worked

### Development
- Updated Rubocop rules
- Fixed specs to run with Chefspec 4.X

v6.0.4
------
### Bug
- Fix normalized hostnames not normalizing the hostgroups
- Don't register the service templates so that Nagios will start properly
- Require Apache2 cookbook version 2.0 or greater due to breaking changes with how site.conf files are handled

### Improvement
- Added additional options for perfdata

### New Feature
- Added the ability to specify a URL to download patches that will be applied to the source install prior to compliation


v6.0.2
------
### Bug
- Remove .DS_Store files in the supermarket file that caused failures on older versions of Berkshelf

v6.0.0
------
### Breaking changes

attributes/default.rb
#
# Author:: Seth Chisamore <schisamo@getchef.com>
# Author:: Tim Smith <tim@cozy.co>
# Cookbook Name:: nagios
# Attributes:: default
#
# Copyright 2011-2013, Chef Software, Inc.
# Copyright 2013-2014, Limelight Networks, Inc
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Allow a Nagios server to monitor hosts in multiple environments.
default['nagios']['multi_environment_monitoring'] = false
default['nagios']['monitored_environments'] = []

default['nagios']['user']  = 'nagios'
default['nagios']['group'] = 'nagios'

# Allow specifying which interface on clients to monitor (which IP address to monitor)
default['nagios']['monitoring_interface'] = nil

case node['platform_family']
when 'debian'
  default['nagios']['plugin_dir'] = '/usr/lib/nagios/plugins'
when 'rhel', 'fedora'
  if node['kernel']['machine'] == 'i686'
    default['nagios']['plugin_dir'] = '/usr/lib/nagios/plugins'
  else
    default['nagios']['plugin_dir'] = '/usr/lib64/nagios/plugins'
  end
else
  default['nagios']['plugin_dir'] = '/usr/lib/nagios/plugins'
end

# platform specific directories
case node['platform_family']
when 'rhel', 'fedora'
  default['nagios']['home']          = '/var/spool/nagios'
  default['nagios']['conf_dir']      = '/etc/nagios'
  default['nagios']['resource_dir']  = '/etc/nagios'
  default['nagios']['config_dir']    = '/etc/nagios/conf.d'
  default['nagios']['log_dir']       = '/var/log/nagios'
  default['nagios']['cache_dir']     = '/var/log/nagios'
  default['nagios']['state_dir']     = '/var/log/nagios'
  default['nagios']['run_dir']       = '/var/run'
  default['nagios']['docroot']       = '/usr/share/nagios/html'
  default['nagios']['cgi-bin']       = '/usr/lib64/nagios/cgi-bin/'
else
  default['nagios']['home']          = '/usr/lib/nagios3'
  default['nagios']['conf_dir']      = '/etc/nagios3'
  default['nagios']['resource_dir']  = '/etc/nagios3'
  default['nagios']['config_dir']    = '/etc/nagios3/conf.d'
  default['nagios']['log_dir']       = '/var/log/nagios3'
  default['nagios']['cache_dir']     = '/var/cache/nagios3'
  default['nagios']['state_dir']     = '/var/lib/nagios3'
  default['nagios']['run_dir']       = '/var/run/nagios3'
  default['nagios']['docroot']       = '/usr/share/nagios3/htdocs'
  default['nagios']['cgi-bin']       = '/usr/lib/cgi-bin/nagios3'
end

# platform specific atttributes
case node['platform_family']
when 'debian'
  default['nagios']['server']['install_method'] = 'package'
  default['nagios']['server']['service_name']   = 'nagios3'
  default['nagios']['server']['mail_command']   = '/usr/bin/mail'
  default['nagios']['cgi-path'] = "/cgi-bin/#{node['nagios']['server']['service_name']}"
when 'rhel', 'fedora'
  default['nagios']['cgi-path'] = '/nagios/cgi-bin/'
  # install via source on RHEL releases less than 6, otherwise use packages
  if node['platform_family'] == 'rhel' && node['platform_version'].to_i < 6
    default['nagios']['server']['install_method'] = 'source'
  else
    default['nagios']['server']['install_method'] = 'package'
  end
  default['nagios']['server']['service_name']   = 'nagios'
  default['nagios']['server']['mail_command']   = '/bin/mail'
else
  default['nagios']['server']['install_method'] = 'source'
  default['nagios']['server']['service_name']   = 'nagios'
  default['nagios']['server']['mail_command']   = '/bin/mail'
end

# webserver configuration
default['nagios']['timezone']      = 'UTC'
default['nagios']['enable_ssl']    = false
default['nagios']['http_port']     = node['nagios']['enable_ssl'] ? '443' : '80'
default['nagios']['server_name']   = node['fqdn']
default['nagios']['server']['server_alias'] = nil
default['nagios']['ssl_cert_file'] = "#{node['nagios']['conf_dir']}/certificates/nagios-server.pem"
default['nagios']['ssl_cert_key']  = "#{node['nagios']['conf_dir']}/certificates/nagios-server.pem"
default['nagios']['ssl_req']       = '/C=US/ST=Several/L=Locality/O=Example/OU=Operations/' \
  "CN=#{node['nagios']['server_name']}/emailAddress=ops@#{node['nagios']['server_name']}"

# nagios server name and webserver vname.  this can be changed to allow for the installation of icinga
default['nagios']['server']['name'] = 'nagios'
case node['platform_family']
when 'rhel', 'fedora'
  default['nagios']['server']['vname'] = 'nagios'
else
  default['nagios']['server']['vname'] = 'nagios3'
end

# for server from source installation
default['nagios']['server']['url']       = 'http://iweb.dl.sourceforge.net/project/nagios/nagios-4.x/nagios-4.0.8/nagios-4.0.8.tar.gz'
default['nagios']['server']['checksum']  = '8b268d250c97851775abe162f46f64724f95f367d752ae4630280cc5d368ca4b'
default['nagios']['server']['src_dir']   = node['nagios']['server']['url'].split('/')[-1].chomp('.tar.gz')
default['nagios']['server']['patches']   = []
default['nagios']['server']['patch_url'] = nil

# for server from packages installation
case node['platform_family']
when 'rhel', 'fedora'
  default['nagios']['server']['packages'] = %w(nagios nagios-plugins-nrpe)
  default['nagios']['server']['install_yum-epel'] = true
else
  default['nagios']['server']['packages'] = %w(nagios3 nagios-nrpe-plugin nagios-images)
end

default['nagios']['check_external_commands']     = true
default['nagios']['default_contact_groups']      = %w(admins)
default['nagios']['sysadmin_email']              = 'root@localhost'
default['nagios']['sysadmin_sms_email']          = 'root@localhost'
default['nagios']['server_auth_method']          = 'htauth'
default['nagios']['server_auth_require']         = 'valid-user'
default['nagios']['users_databag']               = 'users'
default['nagios']['users_databag_group']         = 'sysadmin'
default['nagios']['services_databag']            = 'nagios_services'
default['nagios']['servicegroups_databag']       = 'nagios_servicegroups'
default['nagios']['templates_databag']           = 'nagios_templates'
default['nagios']['hosttemplates_databag']       = 'nagios_hosttemplates'
default['nagios']['eventhandlers_databag']       = 'nagios_eventhandlers'
default['nagios']['unmanagedhosts_databag']      = 'nagios_unmanagedhosts'
default['nagios']['serviceescalations_databag']  = 'nagios_serviceescalations'
default['nagios']['hostgroups_databag']          = 'nagios_hostgroups'
default['nagios']['hostescalations_databag']     = 'nagios_hostescalations'
default['nagios']['contacts_databag']            = 'nagios_contacts'
default['nagios']['contactgroups_databag']       = 'nagios_contactgroups'
default['nagios']['servicedependencies_databag'] = 'nagios_servicedependencies'
default['nagios']['timeperiods_databag']         = 'nagios_timeperiods'
default['nagios']['host_name_attribute']         = 'hostname'
default['nagios']['regexp_matching']             = 0
default['nagios']['large_installation_tweaks']   = 0
default['nagios']['host_template'] = 'server'

# for cas authentication
default['nagios']['cas_login_url']       = 'https://example.com/cas/login'
default['nagios']['cas_validate_url']    = 'https://example.com/cas/serviceValidate'
default['nagios']['cas_validate_server'] = 'off'
default['nagios']['cas_root_proxy_url']  = nil

# for apache ldap authentication
default['nagios']['ldap_bind_dn']       = nil
default['nagios']['ldap_bind_password'] = nil
default['nagios']['ldap_url']           = nil
default['nagios']['ldap_authoritative'] = nil

default['nagios']['templates']       = Mash.new

default['nagios']['nodes_search_query'] = "hostname:* AND chef_environment:#{node.chef_environment}"

# This setting is effectively sets the minimum interval (in seconds) nagios can handle.
# Other interval settings provided in seconds will calculate their actual from this value, since nagios works in 'time units' rather than allowing definitions everywhere in seconds
default['nagios']['interval_length'] = 1

default['nagios']['default_host']['flap_detection']        = true
default['nagios']['default_host']['process_perf_data']     = false
default['nagios']['default_host']['check_period']          = '24x7'
# Provide all interval values in seconds
default['nagios']['default_host']['check_interval']        = 15
default['nagios']['default_host']['retry_interval']        = 15
default['nagios']['default_host']['max_check_attempts']    = 1
default['nagios']['default_host']['check_command']         = 'check_host_alive'
default['nagios']['default_host']['notification_interval'] = 300
default['nagios']['default_host']['notification_options']  = 'd,u,r'
default['nagios']['default_host']['action_url']            = nil

default['nagios']['default_service']['check_interval']        = 60
default['nagios']['default_service']['process_perf_data']     = false
default['nagios']['default_service']['retry_interval']        = 15
default['nagios']['default_service']['max_check_attempts']    = 3
default['nagios']['default_service']['notification_interval'] = 1200
default['nagios']['default_service']['flap_detection']        = true
default['nagios']['default_service']['action_url']            = nil

default['nagios']['server']['web_server']              = 'apache'
default['nagios']['server']['nginx_dispatch']          = 'cgi'
default['nagios']['server']['stop_apache']             = false
default['nagios']['server']['normalize_hostname']      = false
default['nagios']['server']['load_default_config']     = true
default['nagios']['server']['load_databag_config']     = true
default['nagios']['server']['use_encrypted_data_bags'] = false

default['nagios']['cgi']['show_context_help']                        = 1
default['nagios']['cgi']['authorized_for_system_information']        = '*'
default['nagios']['cgi']['authorized_for_configuration_information'] = '*'
default['nagios']['cgi']['authorized_for_system_commands']           = '*'
default['nagios']['cgi']['authorized_for_all_services']              = '*'
default['nagios']['cgi']['authorized_for_all_hosts']                 = '*'
default['nagios']['cgi']['authorized_for_all_service_commands']      = '*'
default['nagios']['cgi']['authorized_for_all_host_commands']         = '*'
default['nagios']['cgi']['default_statusmap_layout']                 = 5
default['nagios']['cgi']['default_statuswrl_layout']                 = 4
default['nagios']['cgi']['escape_html_tags']                         = 0
default['nagios']['cgi']['action_url_target']                        = '_blank'
default['nagios']['cgi']['notes_url_target']                         = '_blank'
default['nagios']['cgi']['lock_author_names']                        = 1

default['nagios']['pagerduty']['script_url'] = 'https://raw.github.com/PagerDuty/pagerduty-nagios-pl/master/pagerduty_nagios.pl'
default['nagios']['pagerduty']['service_notification_options'] = 'w,u,c,r'
default['nagios']['pagerduty']['host_notification_options'] = 'd,r'

# atrributes for setting broker lines
default['nagios']['brokers'] = {}

# attribute defining tag used to exclude hosts
default['nagios']['exclude_tag_host'] = ''

# Set the prefork module for Apache as PHP is not thread-safe
default['apache']['mpm'] = 'prefork'

# attribute to add commands to source build
default['nagios']['source']['add_build_commands'] = ['make install-exfoliation']
